### PR TITLE
Add Nix flake

### PR DESCRIPTION
**Pull request by nothingnesses**
Should address #834.

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -0,0 +1,6 @@
+(import (let lock = builtins.fromJSON (builtins.readFile ./flake.lock);
+in fetchTarball {
+  url =
+    "https://github.com/edolstra/flake-compat/archive/${lock.nodes.flake-compat.locked.rev}.tar.gz";
```
Comment (sassman) [inline default.nix:4]: can you please elaborate on what URL that is, who maintains that, and how can we be sure this is safe to use?

Comment (nothingnesses) [reply default.nix:4]: Sure thing. That's the URL for the flake-compat library which allows projects using the newer Flakes mechanism (i.e., what I'm adding with this PR) [to be used with derivations that don't use Flakes](https://nixos.wiki/wiki/Flakes#Using_flakes_project_from_a_legacy_Nix). It's maintained by Eelco Dolstra who wrote [the thesis that gave birth to the Nix package manager](https://edolstra.github.io/pubs/phd-thesis.pdf). The [source code for the library](https://github.com/edolstra/flake-compat/blob/master/default.nix) is somewhat small (200 lines), so it's feasible to review it quite quickly for anything malicious.

**Comment by neosam** [comment]
I personally think it would make more sense to put those files in a separate repository.  Instead of using `src=./.`, you could use `src=fetchFromGitHub` to check out the cargo-generate's source code.

**Comment by nothingnesses** [comment]
@neosam Ah, I didn't know that was possible. I'll have to try that out, thanks.

**Comment by neosam** [comment]
> @neosam Ah, I didn't know that was possible. I'll have to try that out, thanks.

You're very welcome.  Not sure if you need the latest version of cargo-generate but there is already cargo-generate in Nix. https://search.nixos.org/packages?channel=22.11&show=cargo-generate&from=0&size=50&sort=relevance&type=packages&query=cargo-generate

**Pull request closed without merging**